### ISSUE TITLE
>=eventlet-0.20.0 'import six' imports the bundled deventlet six version

### ISSUE THREAD

**Issue by hwoarang** [eventlet/eventlet#381]
Hi,

We are seeing a very strange problem in one of our CentOS7 boxes

When using eventlet-0.19.0 with 'six' then importing 'six' works a expected

```
python -c 'import eventlet; import six; print eventlet.__version__; print six.__version__'
0.19.0
1.10.0
```

But with >=eventlet-0.20.0 something strange is happening

```
python -c 'import eventlet; import six; print eventlet.__version__; print six.__version__'
0.20.0
1.8.0
```

```
python -c 'import eventlet; import six; print eventlet.__version__; print six.__version__'
0.20.1
1.8.0
```
This is really wrong since 'import six' shouldn't import the bundled six module from eventlet but use the system's 'six' version

**Comment by temoto** [comment]
`python -V`

**Comment by hwoarang** [comment]
```
~$ python -V
Python 2.7.5
```

**Comment by temoto** [comment]
Can you reproduce it on another machine, not CentOS7?

`python=python2.7 bash test-381`  in theory, this script should test for this issue, but it doesn't trigger error for me.

```
#!/bin/bash
set -eux

: ${python:=}
package=${1-"-e ."}
venv=./venv-issue-381

echo "testing with pip install $package" >&2
rm -rf "$venv"
virtualenv --python=$python "$venv"
#$venv/bin/pip install -U pip setuptools
$venv/bin/pip install 'six==1.10.0' $package
six_version=$($venv/bin/python -c 'import eventlet ; import six ; print(six.__version__)')
test "$six_version" = "1.10.0"
```

**Comment by hwoarang** [comment]
No I can't reproduce it on another host. Other CentOS 7 hosts appear to behave correctly. However, the fact that 0.19.0 works OK on that box suggests that something subtle has changed between 0.19 and 0.20 and triggers this behavior.

**Comment by temoto** [comment]
Uhm, yeah, but those two facts together mean that some subtle change combined with something on that particular machine triggers the error. Can you try completely purging all files related to eventlet and six and try again? I stress the point of not relying on any package management smartness.

**Comment by hwoarang** [comment]
Yes I completely removed all eventlet and six files from the host and installed again. It still does not resolve the issue. I even removed the eventlet dependencies and nothing changed. I install both 'six' and 'eventlet' using 'pip' which shouldn't do any weird thing. If I execute this

```
python -c 'import eventlet; import sys; print sys.modules["six"]'
<module 'six' from '/usr/lib/python2.7/site-packages/eventlet/support/six.pyc'>
```
you see that for whatever reason the 'six' symbol is binded to the eventlet six module during the 'eventlet' import.

On the other hand, a regular 'import six' works fine

```
python -c 'import six; import sys; print sys.modules["six"]'
<module 'six' from '/usr/lib/python2.7/site-packages/six.pyc'>
```

Unfortunately I can't really explain or even understand where the root cause of this could be. But since 0.19 works I believe that looking at the differences between 0.19 and 0.20 is a good start.

**Comment by temoto** [comment]
Do you have same Python version on different machines?

**Comment by temoto** [comment]
And what about `setuptools` versions?

**Comment by hwoarang** [comment]
Yes it's same python and setuptools (34.0.2)

**Comment by hwoarang** [comment]
This might also be of interest which is printed when executing

```
python -vvv -c 'import eventlet'

...
# trying /usr/lib/python2.7/site-packages/ecdsa/six.so
# trying /usr/lib/python2.7/site-packages/ecdsa/sixmodule.so
# trying /usr/lib/python2.7/site-packages/ecdsa/six.py
# trying /usr/lib/python2.7/site-packages/ecdsa/six.pyc
# trying /usr/lib/python2.7/site-packages/eventlet/support/six.so
# trying /usr/lib/python2.7/site-packages/eventlet/support/sixmodule.so
# trying /usr/lib/python2.7/site-packages/eventlet/support/six.py
# /usr/lib/python2.7/site-packages/eventlet/support/six.pyc matches /usr/lib/python2.7/site-packages/eventlet/support/six.py
import six # precompiled from /usr/lib/python2.7/site-packages/eventlet/support/six.pyc
...
```

**Comment by hwoarang** [comment]
Sure

broken:

```
python -c 'import sys; print sys.path'
['', '/usr/lib64/python27.zip', '/usr/lib64/python2.7', '/usr/lib64/python2.7/plat-linux2', '/usr/lib64/python2.7/lib-tk', '/usr/lib64/python2.7/lib-old', '/usr/lib64/python2.7/lib-dynload', '/usr/lib64/python2.7/site-packages', '/usr/lib64/python2.7/site-packages/gtk-2.0', '/usr/lib/python2.7/site-packages']
```

working:
```
['', '/usr/lib64/python27.zip', '/usr/lib64/python2.7', '/usr/lib64/python2.7/plat-linux2', '/usr/lib64/python2.7/lib-tk', '/usr/lib64/python2.7/lib-old', '/usr/lib64/python2.7/lib-dynload', '/usr/lib64/python2.7/site-packages', '/usr/lib/python2.7/site-packages']
```

they look pretty similar

**Comment by hwoarang** [comment]
Bisection also let me to this as the first bad commit

```
52b09becacd23f384cf69ae37d70c893c43e3b13 is the first bad commit
commit 52b09becacd23f384cf69ae37d70c893c43e3b13
Author: Sergey Shepelev <temotor@gmail.com>
Date:   Sat Aug 20 19:40:22 2016 +0500

    bundle dnspython in support/; resolving is always green

    Fixes installation issue when older dnspython is present in system packages
    https://github.com/eventlet/eventlet/pull/341
```

**Comment by temoto** [comment]
Yeah I know that commit is the only thing that could broke it, but I thought it's crafted smart enough to not break anything.

Can you also install this thing `site-packages/gtk-2.0` on healthy machine and try again?

**Comment by temoto** [comment]
it's not gonna hurt to add clickable reference to breaking commit and issue: 52b09becacd23f384cf69ae37d70c893c43e3b13 https://github.com/eventlet/eventlet/pull/341

**Comment by hwoarang** [comment]
Sadly installing pygtk2 didn't make any difference. The paths are now the same but nothing really changed

**Comment by temoto** [comment]
Thanks for all your input, I'm out of ideas right now what's going on.

Is it possible you share all python related files, so I try to recreate the conditions in virtual machine and try more thorough step-by-step debugging?

`find /usr -xdev -print0 -path '/usr/*/python*' |tar --null -cf - -T - |xz -c0 >eventlet-issue-381-filesystem.tar.xz`
Note it will copy any globally installed packages from maybe project you work on. You can share it via private torrent or upload to some cloud storage with link-only access.

**Comment by temoto** [comment]
Also please show `pwd ; ls -lh .` before running test script on broken machine.

**Comment by temoto** [comment]
Also, I know this is going insane road but please, try this version

`pip install https://github.com/eventlet/eventlet/archive/wrong-six-381.zip`

**Comment by hwoarang** [comment]
So the `wrong-six` package didn't fix it. Uploading all the python related files could be a bit problematic but I will see if I can do that.

**Comment by temoto** [comment]
By the way, if the real problem is eventlet's bundled six is too old for something else in your system, a temporary crutch workaround is to replace `eventlet/support/six.py` to whatever version you need.

**Comment by hwoarang** [comment]
Yes, I have already replaced the bundled files with the 1.10 ones. I suspect the problem will be "masked" once you make a new release which will include 1.10 anyway

**Comment by temoto** [comment]
@rnurgaliyev 
- #479 is an issue not PR
- https://github.com/eventlet/eventlet/commit/81c7cf94afbb37fcd406cad032b5e73444c6cc10 fixes problem with dnspython
- 53b63c66f0d9a7d6ba9182a207529ebd949e95f3 fixes issue with six by making it external dependency, not bundled in eventlet code
